src/api/crypto/blockchain/BalanceOracle.cpp

```diff
@@ -284,22 +269,18 @@ auto BalanceOracle::Imp::process_update_balance(
     Balance input) noexcept -> void
 {
     auto changed{false};
-    auto& data = [&]() -> auto&
-    {
-        if (auto i = data_.find(chain); i != data_.end()) {
-            auto& out = i->second.first;
-            changed = (out.first != input);
-
-            return out;
-        } else {
-            changed = true;
-
-            return data_[chain].first;
-        }
+    Data* data;
```

Comment:
As a minimum, Data* data{}; 
or explicitly data{nullptr} if you prefer. This is still convoluted for no good reason, so better:
if (auto i = data_.find(chain); i != data_.end()) {
  auto& balance = i->second.first.first;
  balance = std::move(input);
} else {
  auto& balance = data_[chain].first.first;
  balance = std::move(input);
  auto subscribers = data_[chain].first.second;
  notify_subscribers(subscribers,balance,chain);
}
The second option is clearer and easier to trace. It requires a double check that "Balance input" is a temporary value, in which case a swap is wasteful overkill replacement for a one-way move.
If you were inclined to use the "long" option, please double check the correctness.

Reply:
If that is not a problem I would leave version with pointer. I will initialize it, add const Balance& input as parameter (no idea why it was copied), and change swap to move